Add best practices page to Dask cuDF docs (#16821)
rapidsai/cudf: docs/dask_cudf/source/best_practices.rst

.. _best-practices:

Dask cuDF Best Practices
========================

This page outlines several important guidelines for using Dask cuDF
effectively.

.. note::
  Since Dask cuDF is a backend extension for
  `Dask DataFrame <https://docs.dask.org/en/stable/dataframe.html>`__,
  many of the details discussed in the `Dask DataFrames Best Practices
  <https://docs.dask.org/en/stable/dataframe-best-practices.html>`__

>   documentation also apply to Dask cuDF.

Author: I updated the wording to say the guidelines that are not pandas-specific also apply to Dask cuDF.

Deployment and Configuration
----------------------------

> Use Dask DataFrame Directly

Author: Not sure what "section" this belongs to.

  <https://docs.dask.org/en/latest/deploying-kubernetes.html>`__ and `Dask-Jobqueue
  <https://jobqueue.dask.org/en/latest/>`__.

  Please see `RAPIDS-deployment documentation <https://docs.rapids.ai/deployment/stable/>`__
  for further details and examples.

Enable cuDF Spilling
~~~~~~~~~~~~~~~~~~~~

When using Dask cuDF for classic ETL workloads, it is usually best
to enable `native spilling support in cuDF
<https://docs.rapids.ai/api/cudf/stable/developer_guide/library_design/#spilling-to-host-memory>`__.
When using :func:`LocalCUDACluster`, this is easily accomplished by
setting ``enable_cudf_spill=True``.

When a Dask cuDF workflow includes conversion between DataFrame and Array
representations, native cuDF spilling may be insufficient. For these cases,
JIT unspilling is likely to produce better protection from out-of-memory
(OOM) errors. Please see `Dask-CUDA's spilling documentation
<https://docs.rapids.ai/api/dask-cuda/24.10/spilling/>`__ for further details
and guidance.

Reading Data
------------

Tune the partition size
~~~~~~~~~~~~~~~~~~~~~~~

The ideal partition size is typically between 2-10% of the memory capacity

> of a single GPU. Increasing the partition size will typically reduce the

Author: Okay, I attempted to turn this into the more-explicit "rule of thumb" I personally use: 1/16 or less if the workflow is memory-intensive (i.e. shuffle intensive), and 1/8 otherwise. The "best" partition size is definitely difficult to know a priori.

task. However, if the partitions are too large, the risk of OOM errors can
become significant.

The best way to tune the partition size is to begin with appropriate sized
partitions when the DataFrame collection is first created by a function
like :func:`read_parquet`, :func:`read_csv`, or :func:`from_map`. For
example, both :func:`read_parquet` and :func:`read_csv` expose a
``blocksize`` argument for adjusting the maximum partition size.

If the partition size cannot be tuned effectively at creation time, the
`repartition <https://docs.dask.org/en/latest/generated/dask.dataframe.DataFrame.repartition.html>`__
method can be used as a last resort.

Use Parquet files
~~~~~~~~~~~~~~~~~

`Parquet <https://parquet.apache.org/docs/file-format/>`__ is the recommended
file format for Dask cuDF. It provides efficient columnar storage and enables
Dask to perform valuable query optimizations like column projection and
predicate pushdown.

The most important arguments to :func:`read_parquet` are ``blocksize`` and
``aggregate_files``::

``blocksize``: Use this argument to specify the maximum partition size.
The default is `"256 MiB"`, but larger values are usually more performant

> (e.g. `1 GiB` is usually safe). Dask will use the ``blocksize`` value to map

Author: I decided to remove the 1 GiB comment since we already discuss the 1/16-1/8 "rule of thumb" above.

contains many files that are smaller than half of ``blocksize``.

.. note::
  Metadata collection can be extremely slow when reading from remote
  storage (e.g. S3 and GCS). When reading many remote files that all
  correspond to a reasonable partition size, it's usually best to set
  `blocksize=None` and `aggregate_files=False`. In most cases, these
  settings allow Dask to skip the metadata-collection stage altogether.

Use :func:`from_map`
~~~~~~~~~~~~~~~~~~~~

To implement custom DataFrame-creation logic that is not covered by
existing APIs (like :func:`read_parquet`),  use :func:`dask.dataframe.from_map`
whenever possible. The :func:`from_map` API has several advantages
over :func:`from_delayed`::

* It allows proper lazy execution of your custom logic
* It enables column projection (as long as the mapped function supports a ``columns`` key-word argument)

See the `from_map API documentation <https://docs.dask.org/en/stable/generated/dask_expr.from_map.html#dask_expr.from_map>`__
for more details.

Sorting, Joining and Grouping
-----------------------------

Sorting, joining and grouping operations all have the potential to

> require the global shuffling of data between distinct partitions.

Author: Hmmm. Repartition does not really require data "shuffling". Data shuffling requires "all-to-all", while repartitioning is usually limited to data movement between neighboring partitions.

When the initial data fits comfortably in global GPU memory, these
"all-to-all" operations are typically bound by worker-to-worker
communication. When the data is larger than global GPU memory, the
bottleneck is typically device-to-host memory spilling.

Although every workflow is different, the following guidelines
are often recommended::

* Use a distributed cluster with Dask-CUDA workers

> * Use native cuDF spilling whenever possible

Reviewer: Should we link to the sections above in here that deal with these two suggestions, for the benefit of the reader who may come directly to this section? Not sure if easy/possible, if not skipping is also fine.

Author: Yeah, good idea. Will need to figure out how to do that :)

When using :func:`LocalCUDACluster`, this is easily accomplished by
setting ``enable_cudf_spill=True``.

When a Dask cuDF workflow includes conversion between DataFrame and Array
representations, native cuDF spilling may be insufficient. For these cases,
JIT unspilling is likely to produce better protection from out-of-memory
(OOM) errors. Please see `Dask-CUDA's spilling documentation
<https://docs.rapids.ai/api/dask-cuda/24.10/spilling/>`__ for further details
and guidance.

Use RMM
~~~~~~~

Memory allocations in cuDF are significantly faster and more efficient when
the `RAPIDS Memory Manager (RMM) <https://docs.rapids.ai/api/rmm/stable/>`__
library is used on worker processes. In most cases, the best way to manage
memory is by initializing an RMM pool on each worker before executing a
workflow. When using :func:`LocalCUDACluster`, this is easily accomplished
by setting ``rmm_pool_size`` to a large fraction (e.g. ``0.9``).

See the `Dask-CUDA memory-management documentation
<https://docs.rapids.ai/api/dask-cuda/nightly/examples/best-practices/#gpu-memory-management>`__
for more details.

Use the Dask DataFrame API

> ~~~~~~~~~~~~~~~~~~~~~~~~~~

Reviewer 2: It's probably a repeat of Dask's best practices but might be worth calling out that users should try to avoid `compute` when possible (unless results are extremely small) especially in a GPU setting where memory is limited. Similarly `persist` only if the results comfortable fit in GPU memory.

Although Dask cuDF provides a public ``dask_cudf`` Python module, we
strongly recommended that you use the CPU/GPU portable ``dask.dataframe``
API instead. Simply use the `Dask configuration <dask:configuration>`__
system to set the ``"dataframe.backend"`` option to ``"cudf"``, and

> the ``dask_cudf`` module will be imported and used implicitly.

Reviewer 2: Nit: Not sure if the link to `Dask:configuration` works here or not.

Additionally,  the Dask config page has a lot of information not specific to the backend so I prefer also sharing a small snippet or linking to an example that does this to illustrate the point.

Author: Okay yeah, I switched to the explicit url to be safe.

errors can become significant.

.. note::
  As a general rule of thumb, aim for 1/16 in shuffle-intensive workflows
  (e.g. large-scale sorting and joining), and 1/8 otherwise. For pathologically
  skewed data distributions, it may be necessary to target 1/32 or smaller.

The easiest way to tune the partition size is when the DataFrame collection
is first created by a function like :func:`read_parquet`, :func:`read_csv`,
or :func:`from_map`. For example, both :func:`read_parquet` and :func:`read_csv`
expose a ``blocksize`` argument for adjusting the maximum partition size.

If the partition size cannot be tuned effectively at creation time, the
`repartition <https://docs.dask.org/en/latest/generated/dask.dataframe.DataFrame.repartition.html>`__
method can be used as a last resort.

Use Parquet
~~~~~~~~~~~

`Parquet <https://parquet.apache.org/docs/file-format/>`__ is the recommended
file format for Dask cuDF. It provides efficient columnar storage and enables
Dask to perform valuable query optimizations like column projection and
predicate pushdown.

The most important arguments to :func:`read_parquet` are ``blocksize`` and

> ``aggregate_files``::

Reviewer 2: Not sure if it's super important, but something I've come across (might be super specific to my workflows): If the data is already pre shuffled or sorted along some attribute, reading with a smaller block sizes might lead to incorrect partitioning. I tend to use `split_row_groups=False`  (or the equivalent block size) for those cases.

Author: Okay yeah, I actually added a note that you should use `from_map` in any case that you must have 1-to-1 mapping between files and partitions. Hopefully it is clear enough from the docs that `blocksize=None` will avoid breaking files into multiple partitions?

Small note: I don't want `split_row_groups` to show up anywhere here, because that option is essentially obsolete now (and may be deprecated).

Reviewer 2: That makes sense, perhaps `blocksize=None` is the best replacement to get the same point across.

``blocksize``: Use this argument to specify the maximum partition size.
The default is `"256 MiB"`, but larger values are usually more performant
on GPUs with more than 8 GiB of memory. Dask will use the ``blocksize``
value to map a discrete number of Parquet row-groups (or files) to each
output partition. This mapping will only account for the uncompressed
storage size of each row group, which is usually smaller than the
correspondng ``cudf.DataFrame``.

``aggregate_files``: Use this argument to specify whether Dask is allowed
to map multiple files to the same DataFrame partition. The default is
``False``, but ``aggregate_files=True`` is usually more performant when
the dataset contains many files that are smaller than half of ``blocksize``.

> .. note::

Reviewer 2: I like this note. Once we have more cloud IO specific optimizations it might make sense to add it to best practices or create a new one for cloud IO to discuss tips/tricks for those environments.

Author: Agree that we need a lot more remote-IO information. However, it doesn't feel like there is much to say yet :/

  Metadata collection can be extremely slow when reading from remote
  storage (e.g. S3 and GCS). When reading many remote files that all
  correspond to a reasonable partition size, it's usually best to set
  `blocksize=None` and `aggregate_files=False`. In most cases, these
  settings allow Dask to skip the metadata-collection stage altogether.

Use :func:`from_map`
~~~~~~~~~~~~~~~~~~~~

To implement custom DataFrame-creation logic that is not covered by
existing APIs (like :func:`read_parquet`),  use :func:`dask.dataframe.from_map`
whenever possible. The :func:`from_map` API has several advantages
over :func:`from_delayed`::

* It allows proper lazy execution of your custom logic
* It enables column projection (as long as the mapped function supports a ``columns`` key-word argument)

> See the `from_map API documentation <https://docs.dask.org/en/stable/generated/dask_expr.from_map.html#dask_expr.from_map>`__

Reviewer 2: One question around `from_map`. In my experience using `from_map` for meta collection involves reading in a single partition which can often have high memory overhead (on the client process) when dealing with larger partitions and therefore conflicting with a large rmm pool size of `0.9`. It might be worth recommending either skipping the single partition meta calculation somehow, or suggestion using a smaller pool size if users run into similar issues.

Author: Yeah, this is a good point. It's definitely important to specify `meta` when `from_map` is used! If its just using `cudf.read_parquet`, you can usually avoid the RMM problem by doing something like: `meta=cudf.read_parquet(..., row_groups=0)`

for more details.

Sorting, Joining and Grouping
-----------------------------

Sorting, joining and grouping operations all have the potential to
require the global shuffling of data between distinct partitions.
When the initial data fits comfortably in global GPU memory, these
"all-to-all" operations are typically bound by worker-to-worker
communication. When the data is larger than global GPU memory, the
bottleneck is typically device-to-host memory spilling.

Although every workflow is different, the following guidelines
are often recommended::

* `Use a distributed cluster with Dask-CUDA workers <Use Dask-CUDA>`_
* `Use native cuDF spilling whenever possible <Enable cuDF Spilling>`_
* Avoid shuffling whenever possible
  * Use ``split_out=1`` for low-cardinality groupby aggregations
  * Use ``broadcast=True`` for joins when at least one collection comprises a small number of partitions (e.g. ``<=5``)

> * `Use UCX <https://docs.rapids.ai/api/dask-cuda/nightly/examples/ucx/>`__ if communication is a bottleneck

Reviewer 2: Nit: It might be also worth mentioning here that UCX is key that allows dask to uses faster transports like `NVLink` `nvswitch` or `inifiband` for better comms

> 

Reviewer 2: In my experience, `shuffle` is often more performant/flexible than `sorting` or `set_index` (before dask-expr, not sure if that's changed now). For cases where shuffling data is not avoidable, considering whether a non sorted `shuffle(ignore_index=True)`  call is sufficient can be better than calling `set_index`.

> 

Reviewer 2: It might be worth adding a section or linking to Dask performance report pages and say if users want to profile workflows they can use the performance report feature.

.. _best-practices:

Dask cuDF Best Practices
========================

This page outlines several important guidelines for using Dask cuDF
effectively.

.. note::
  Since Dask cuDF is a backend extension for
  `Dask DataFrame <https://docs.dask.org/en/stable/dataframe.html>`__,
  the guidelines discussed in the `Dask DataFrames Best Practices
  <https://docs.dask.org/en/stable/dataframe-best-practices.html>`__
  documentation also apply to Dask cuDF (excluding any pandas-specific
  details).

Deployment and Configuration
----------------------------

Use Dask-CUDA
~~~~~~~~~~~~~

In order to execute a Dask workflow on multiple GPUs, a Dask cluster must
be deployed with `Dask-CUDA <https://docs.rapids.ai/api/dask-cuda/stable/>`__

> and/or `Dask.distributed <https://distributed.dask.org/en/stable/>`__.

Author: Okay, I agree - will change to "and". (I think I copied this from previous wording that was a bit more CPU/GPU general)

  <https://docs.dask.org/en/latest/deploying-kubernetes.html>`__ and `Dask-Jobqueue
  <https://jobqueue.dask.org/en/latest/>`__.

  Please see `RAPIDS-deployment documentation <https://docs.rapids.ai/deployment/stable/>`__
  for further details and examples.

Use diagnostic tools
~~~~~~~~~~~~~~~~~~~~

The Dask ecosystem includes several diagnostic tools that you should absolutely use.
These tools include an intuitive `browser dashboard
<https://docs.dask.org/en/stable/dashboard.html>`__ as well as a dedicated
`API for collecting performance profiles
<https://distributed.dask.org/en/latest/diagnosing-performance.html#performance-reports>`__.

Enable cuDF spilling
~~~~~~~~~~~~~~~~~~~~

When using Dask cuDF for classic ETL workloads, it is usually best
to enable `native spilling support in cuDF
<https://docs.rapids.ai/api/cudf/stable/developer_guide/library_design/#spilling-to-host-memory>`__.
When using :func:`LocalCUDACluster`, this is easily accomplished by
setting ``enable_cudf_spill=True``.

When a Dask cuDF workflow includes conversion between DataFrame and Array
representations, native cuDF spilling may be insufficient. For these cases,

> JIT unspilling is likely to produce better protection from out-of-memory

Author: I linked to the docs on JIT-unspill instead. That seem okay?

of the underlying task graph to materialize the collection.

:func:`sort_values` / :func:`set_index` : These operations both require Dask to
eagerly collect quantile information about the column(s) being targeted by the
global sort operation. See `Avoid Sorting`__ for notes on sorting considerations.

.. note::
  When using :func:`set_index`, be sure to pass in ``sort=False`` whenever the
  global collection does not **need** to be sorted by the new index.

Avoid Sorting
~~~~~~~~~~~~~

`The design of Dask DataFrame <https://docs.dask.org/en/stable/dataframe-design.html#dask-dataframe-design>`__
makes it advantageous to work with data that is already sorted along its index at
creation time. For most other cases, it is best to avoid sorting unless the logic
of the workflow makes global ordering absolutely necessary.

If the purpose of a :func:`sort_values` operation is to ensure that all unique
values in ``by`` will be moved to the same output partition, then `shuffle
<https://docs.dask.org/en/stable/generated/dask.dataframe.DataFrame.shuffle.html>`__
is often the better option.

Reading Data
------------

Tune the partition size
~~~~~~~~~~~~~~~~~~~~~~~

> The ideal partition size is usually between 1/16 and 1/8 the memory

Author: I added the following to the note below:

"This rule of thumb comes from anecdotal optimization and OOM-debugging experience. Since every workflow is different, choosing the best partition size is both an art and a science."

Does this seem okay?

Additional background: The default in NVTabular/Merlin is to target partitions that are 12.5% single-GPU memory (i.e. 1/8). These workflows usually include tree reductions and avoid global shuffling. We chose that number, because it was optimal for a categorical encoding benchmark, and seemed safe for other workflows as well. For shuffle-heavy workloads, 1/32 - 1/16 seems more common (but these specific numbers are anecdotal).

> The ideal partition size is usually between 1/16 and 1/8 the memory

Reviewer 3: Sounds great! Thanks @rjzamora

Enable cuDF spilling
~~~~~~~~~~~~~~~~~~~~

When using Dask cuDF for classic ETL workloads, it is usually best
to enable `native spilling support in cuDF
<https://docs.rapids.ai/api/cudf/stable/developer_guide/library_design/#spilling-to-host-memory>`__.
When using :func:`LocalCUDACluster`, this is easily accomplished by
setting ``enable_cudf_spill=True``.

When a Dask cuDF workflow includes conversion between DataFrame and Array
representations, native cuDF spilling may be insufficient. For these cases,
`JIT-unspill <https://docs.rapids.ai/api/dask-cuda/nightly/spilling/#jit-unspill>`__
is likely to produce better protection from out-of-memory (OOM) errors.
Please see `Dask-CUDA's spilling documentation
<https://docs.rapids.ai/api/dask-cuda/24.10/spilling/>`__ for further details
and guidance.

Use RMM
~~~~~~~

Memory allocations in cuDF are significantly faster and more efficient when
the `RAPIDS Memory Manager (RMM) <https://docs.rapids.ai/api/rmm/stable/>`__
library is used on worker processes. In most cases, the best way to manage
memory is by initializing an RMM pool on each worker before executing a
workflow. When using :func:`LocalCUDACluster`, this is easily accomplished

> by setting ``rmm_pool_size`` to a large fraction (e.g. ``0.9``).

Reviewer 4: Related, i think we should add a section soon using ``rmm_pool` plus `dask` plus `pytorch`. Have seen a few customers struggle here so we should probably have a section around that.

Author: Okay yeah. This starts to get a bit "advanced" for this particular section. I think we probably need a distinct page on  interfacing with other GPU libraries. That section would definitely need to discuss details like this.

TBH: I don't know if I'm personally up-to-date on best practices for pytorch integrations :)

<https://docs.rapids.ai/api/dask-cuda/nightly/examples/best-practices/#gpu-memory-management>`__
for more details.

Use the Dask DataFrame API
~~~~~~~~~~~~~~~~~~~~~~~~~~

Although Dask cuDF provides a public ``dask_cudf`` Python module, we
strongly recommended that you use the CPU/GPU portable ``dask.dataframe``
API instead. Simply `use the Dask configuration system
<https://docs.dask.org/en/stable/how-to/selecting-the-collection-backend.html>`__
to set the ``"dataframe.backend"`` option to ``"cudf"``, and the
``dask_cudf`` module will be imported and used implicitly.

Avoid eager execution
~~~~~~~~~~~~~~~~~~~~~

Although Dask DataFrame collections are lazy by default, there are several
notable methods that will result in the immediate execution of the
underlying task graph::

:func:`compute`: Calling ``ddf.compute()`` will materialize the result of
``ddf`` and return a single cuDF object. This is done by executing the entire
task graph associated with ``ddf`` and concatenating its partitions in
local memory on the client process.

.. note::
  Never call :func:`compute` on a large collection that cannot fit comfortably
  in the memory of a single GPU!

> :func:`persist`: Like :func:`compute`, calling ``ddf.persist()`` will
> execute the entire task graph associated with ``ddf``. The important difference
> is that the computed partitions will remain in distributed worker memory instead
> of being concatenated together on the client process.

Reviewer 4: I wonder if its worth calling out that `persist` is `non-blocking` and we recommend using `wait` if we need the blocking syncronization based behavior.

```suggestion
:func:`persist`: Like :func:`compute`, calling ``ddf.persist()`` will
execute the entire task graph associated with ``ddf``. The important difference
is that the computed partitions will remain in distributed worker memory instead
of being concatenated together on the client process. It's important to note that `persist` is non-blocking; for blocking synchronization, consider using `wait`. 
For example, on `ddf = ddf.persist()`, `wait(ddf)` can be used to ensure the task graph is fully executed and persisted before proceeding.
```

Author: Good point!  I expanded this section a bit (along the lines of your suggestion).

require the global shuffling of data between distinct partitions.
When the initial data fits comfortably in global GPU memory, these
"all-to-all" operations are typically bound by worker-to-worker
communication. When the data is larger than global GPU memory, the
bottleneck is typically device-to-host memory spilling.

Although every workflow is different, the following guidelines
are often recommended::

* `Use a distributed cluster with Dask-CUDA workers <Use Dask-CUDA>`_
* `Use native cuDF spilling whenever possible <Enable cuDF Spilling>`_
* Avoid shuffling whenever possible
  * Use ``split_out=1`` for low-cardinality groupby aggregations
  * Use ``broadcast=True`` for joins when at least one collection comprises a small number of partitions (e.g. ``<=5``)
* `Use UCX <https://docs.rapids.ai/api/dask-cuda/nightly/examples/ucx/>`__ if communication is a bottleneck.

.. note::
  UCX enables Dask-CUDA workers to communicate using high-performance
  tansport technologies like `NVLink <https://www.nvidia.com/en-us/data-center/nvlink/>`__
  and Infiniband. Without UCX, inter-process communication will rely
  on TCP sockets.

> 

Reviewer 4: I wonder we should add a section for avoiding CPU-GPU transfers. 

Like I have seen folks go back and forth b/w CPU and GPU  more often than needed, we should probably add a note about that. 

+1. Minimize CPU-GPU data transfers:
+   Whenever possible, keep data on the GPU throughout your workflow. Each transfer
+   between CPU and GPU memory incurs a performance cost.

Author: This is a good point. I added a note to the "Use the Dask DataFrame API" about avoiding unnecessary conversion to the pandas backend. Let me know if you think an entire section/subsection seems necessary.

.. _best-practices:

Dask cuDF Best Practices
========================

This page outlines several important guidelines for using `Dask cuDF
<https://docs.rapids.ai/api/dask-cudf/stable/>`__ effectively.

.. note::
  Since Dask cuDF is a backend extension for
  `Dask DataFrame <https://docs.dask.org/en/stable/dataframe.html>`__,
  the guidelines discussed in the `Dask DataFrames Best Practices
  <https://docs.dask.org/en/stable/dataframe-best-practices.html>`__
  documentation also apply to Dask cuDF (excluding any pandas-specific
  details).

Deployment and Configuration
----------------------------

Use Dask-CUDA
~~~~~~~~~~~~~

> In order to execute a Dask workflow on multiple GPUs, a Dask cluster must

Reviewer 5: ```suggestion
To execute a Dask workflow on multiple GPUs, a Dask cluster must
```

convenience function is strongly recommended. No matter how many GPUs are
available on the machine (even one!), using `Dask-CUDA has many advantages
<https://docs.rapids.ai/api/dask-cuda/stable/#motivation>`__
over default (threaded) execution. Just to list a few:

* Dask-CUDA makes it easy to pin workers to specific devices.
* Dask-CUDA makes it easy to configure memory-spilling options.
* The distributed scheduler collects useful diagnostic information that can be viewed on a dashboard in real time.

Please see `Dask-CUDA's API <https://docs.rapids.ai/api/dask-cuda/stable/>`__
and `Best Practices <https://docs.rapids.ai/api/dask-cuda/stable/examples/best-practices/>`__
documentation for detailed information. Typical ``LocalCUDACluster`` usage
is also illustrated within the multi-GPU section of `Dask cuDF's
<https://docs.rapids.ai/api/dask-cudf/stable/>`__ documentation.

.. note::
  When running on cloud infrastructure or HPC systems, it is usually best to
  leverage system-specific deployment libraries like `Dask Operator
  <https://docs.dask.org/en/latest/deploying-kubernetes.html>`__ and `Dask-Jobqueue
  <https://jobqueue.dask.org/en/latest/>`__.

>   Please see `RAPIDS-deployment documentation <https://docs.rapids.ai/deployment/stable/>`__

Reviewer 5: ```suggestion
  Please see `the RAPIDS deployment documentation <https://docs.rapids.ai/deployment/stable/>`__
```

  for further details and examples.

Use diagnostic tools
~~~~~~~~~~~~~~~~~~~~

The Dask ecosystem includes several diagnostic tools that you should absolutely use.
These tools include an intuitive `browser dashboard
<https://docs.dask.org/en/stable/dashboard.html>`__ as well as a dedicated
`API for collecting performance profiles
<https://distributed.dask.org/en/latest/diagnosing-performance.html#performance-reports>`__.

No matter the workflow, using the dashboard is strongly recommended.
It provides a visual representation of the worker resources and compute
progress. It also shows basic GPU memory and utilization metrics (under

> the ``GPU`` tab). In order to visualize further GPU metrics in JupyterLab,

Reviewer 5: ```suggestion
the ``GPU`` tab). To visualize more detailed GPU metrics in JupyterLab,
```

use `NVDashboard <https://github.com/rapidsai/jupyterlab-nvdashboard>`__.

Enable cuDF spilling
~~~~~~~~~~~~~~~~~~~~

When using Dask cuDF for classic ETL workloads, it is usually best
to enable `native spilling support in cuDF
<https://docs.rapids.ai/api/cudf/stable/developer_guide/library_design/#spilling-to-host-memory>`__.
When using :func:`LocalCUDACluster`, this is easily accomplished by
setting ``enable_cudf_spill=True``.

When a Dask cuDF workflow includes conversion between DataFrame and Array
representations, native cuDF spilling may be insufficient. For these cases,
`JIT-unspill <https://docs.rapids.ai/api/dask-cuda/nightly/spilling/#jit-unspill>`__
is likely to produce better protection from out-of-memory (OOM) errors.
Please see `Dask-CUDA's spilling documentation
<https://docs.rapids.ai/api/dask-cuda/24.10/spilling/>`__ for further details
and guidance.

Use RMM
~~~~~~~

> Memory allocations in cuDF are significantly faster and more efficient when
> the `RAPIDS Memory Manager (RMM) <https://docs.rapids.ai/api/rmm/stable/>`__
> library is used on worker processes. In most cases, the best way to manage

Reviewer 5: ```suggestion
Memory allocations in cuDF are significantly faster and more efficient when
the `RAPIDS Memory Manager (RMM) <https://docs.rapids.ai/api/rmm/stable/>`__
library is configured appropriately on worker processes. In most cases, the best way to manage
```

RMM is always used for allocations, but we need to remember to use a pool

memory is by initializing an RMM pool on each worker before executing a
workflow. When using :func:`LocalCUDACluster`, this is easily accomplished
by setting ``rmm_pool_size`` to a large fraction (e.g. ``0.9``).

See the `Dask-CUDA memory-management documentation
<https://docs.rapids.ai/api/dask-cuda/nightly/examples/best-practices/#gpu-memory-management>`__
for more details.

Use the Dask DataFrame API
~~~~~~~~~~~~~~~~~~~~~~~~~~

Although Dask cuDF provides a public ``dask_cudf`` Python module, we
strongly recommended that you use the CPU/GPU portable ``dask.dataframe``
API instead. Simply `use the Dask configuration system
<https://docs.dask.org/en/stable/how-to/selecting-the-collection-backend.html>`__
to set the ``"dataframe.backend"`` option to ``"cudf"``, and the
``dask_cudf`` module will be imported and used implicitly.

Be sure to use the :func:`to_backend` method if you need to convert
between the different DataFrame backends. For example::

  df = df.to_backend("pandas")  # This gives us a pandas-backed collection

.. note::
  Although :func:`to_backend` makes it easy to move data between pandas
  and cuDF, repetitive CPU-GPU data movement can degrade performance

>   significantly. For optimal results, keep your data on the GPU as often
>   as possible.

Reviewer 5: ```suggestion
  significantly. For optimal results, keep your data on the GPU as much
  as possible.
```

.. note::
  Metadata collection can be extremely slow when reading from remote
  storage (e.g. S3 and GCS). When reading many remote files that all
  correspond to a reasonable partition size, use ``blocksize=None``
  to avoid unnecessary metadata collection.

Use :func:`from_map`
~~~~~~~~~~~~~~~~~~~~

To implement custom DataFrame-creation logic that is not covered by
existing APIs (like :func:`read_parquet`), use :func:`dask.dataframe.from_map`
whenever possible. The :func:`from_map` API has several advantages
over :func:`from_delayed`:

* It allows proper lazy execution of your custom logic
* It enables column projection (as long as the mapped function supports a ``columns`` key-word argument)

See the `from_map API documentation <https://docs.dask.org/en/stable/generated/dask_expr.from_map.html#dask_expr.from_map>`__
for more details.

.. note::
  Whenever possible, be sure to specify the ``meta`` argument to
  :func:`from_map`. If this argument is excluded, Dask will need to
  materialize the first partition eagerly. If a large RMM pool is in
  use on the first visible device, this eager execution on the client
  may lead to an OOM error.

> Sorting, Joining and Grouping
> -----------------------------

Reviewer 5: ```suggestion
Sorting, Joining, and Grouping
------------------------------
```

> Sorting, joining and grouping operations all have the potential to

Reviewer 5: ```suggestion
Sorting, joining, and grouping operations all have the potential to
```

of the underlying task graph to materialize the collection.

:func:`sort_values` / :func:`set_index` : These operations both require Dask to
eagerly collect quantile information about the column(s) being targeted by the
global sort operation. See `Avoid Sorting`__ for notes on sorting considerations.

.. note::
  When using :func:`set_index`, be sure to pass in ``sort=False`` whenever the
  global collection does not **need** to be sorted by the new index.

Avoid Sorting
~~~~~~~~~~~~~

`The design of Dask DataFrame <https://docs.dask.org/en/stable/dataframe-design.html#dask-dataframe-design>`__
makes it advantageous to work with data that is already sorted along its index at
creation time. For most other cases, it is best to avoid sorting unless the logic
of the workflow makes global ordering absolutely necessary.

If the purpose of a :func:`sort_values` operation is to ensure that all unique
values in ``by`` will be moved to the same output partition, then `shuffle
<https://docs.dask.org/en/stable/generated/dask.dataframe.DataFrame.shuffle.html>`__
is often the better option.

Reading Data
------------

Tune the partition size
~~~~~~~~~~~~~~~~~~~~~~~

> The ideal partition size is usually between 1/16 and 1/8 the memory

Author: ```suggestion
The ideal partition size is usually between 1/32 and 1/8 the memory
```

capacity of a single GPU. Increasing the partition size will typically
reduce the number of tasks in your workflow and improve the GPU utilization
for each task. However, if the partitions are too large, the risk of OOM
errors can become significant.

.. note::

>   As a general rule of thumb, start with 1/32-1/16 in shuffle-intensive workflows

Author: ```suggestion
  As a general rule of thumb, start with 1/32-1/16 for shuffle-intensive workflows
```

>   As a general rule of thumb, aim for 1/16 in shuffle-intensive workflows
>   (e.g. large-scale sorting and joining), and 1/8 otherwise. For pathologically
>   skewed data distributions, it may be necessary to target 1/32 or smaller.

Author: ```suggestion
  As a general rule of thumb, start with 1/32-1/16 in shuffle-intensive workflows
  (e.g. large-scale sorting and joining), and 1/16-1/8 otherwise. For pathologically
  skewed data distributions, it may be necessary to target 1/64 or smaller.
```
